### PR TITLE
vtcombo: close query service on drop database

### PR DESCRIPTION
**Pull request by brendar**
## Description

Closes the query service on `DROP DATABASE`, which will close connection pools.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/16605

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by mattlord — COMMENTED** [review]
Thanks, @brendar ! This change makes sense to me there:
```
err := tablet.qsc.QueryService().Close(ctx)
```

We should, however, add some kind of test that demonstrates the problem on main and in turn verifies that it's been fixed in the PR branch.

**Comment by brendar** [comment]
@mattlord I've added assertions to the vtcombo recreate test. It fails on main:
```
    recreate_test.go:119:
        	Error Trace:	/Users/brendan/src/github.com/Shopify/vitess/go/test/endtoend/vtcombo/recreate/recreate_test.go:119
        	Error:      	Max difference between 16 and 27 allowed is 5, but difference was -11
        	Test:       	TestDropAndRecreateWithSameShards
        	Messages:   	not within allowable delta: mysqlConnCountBefore=16, mysqlConnCountAfter=27
--- FAIL: TestDropAndRecreateWithSameShards (1.87s)
```

When testing the fix locally I sometimes saw a difference of 1 connection before and after, which I believe is due to some short lived or lazily initialized connection, so I gave it what I think is a pretty wide delta (5 connections).

The alternative to counting connections would be to parse the `*PoolMaxCap` stats from `/debug/vars` which felt a bit convoluted but might avoid any flakiness.

File: go/test/endtoend/vtcombo/recreate/recreate_test.go [hunk 1]
```diff
@@ -101,13 +101,36 @@ func TestDropAndRecreateWithSameShards(t *testing.T) {
...
 	_, err = cur.Execute(ctx, "DROP DATABASE "+ks1, nil)
 	require.Nil(t, err)
 
 	_, err = cur.Execute(ctx, "CREATE DATABASE "+ks1, nil)
 	require.Nil(t, err)
 
 	assertTabletsPresent(t)
+
+	mysqlConnCountAfter, err := getMySQLConnectionCount(ctx, cur)
+	require.Nil(t, err)
```
Comment (mattlord) [inline go/test/endtoend/vtcombo/recreate/recreate_test.go:116]: Why not test it after the `DROP` and before the `CREATE`? Isn't that where the problem was?

File: go/test/endtoend/vtcombo/recreate/recreate_test.go [hunk 2]
```diff
@@ -101,13 +101,36 @@ func TestDropAndRecreateWithSameShards(t *testing.T) {
...
+
+	// Assert that we're not leaking mysql connections, but allow for some wiggle room due to transient connections
+	assert.InDelta(t, mysqlConnCountBefore, mysqlConnCountAfter, 5,
+		"not within allowable delta: mysqlConnCountBefore=%d, mysqlConnCountAfter=%d", mysqlConnCountBefore, mysqlConnCountAfter)
+}
+
+func getMySQLConnectionCount(ctx context.Context, session *vtgateconn.VTGateSession) (int, error) {
+	result, err := session.Execute(ctx, "SELECT COUNT(*) FROM information_schema.processlist", nil)
```
Comment (mattlord) [inline go/test/endtoend/vtcombo/recreate/recreate_test.go:124]: IMO this is a better query as the process list will show internal threads too:
```
mysql> show processlist;
+----+-----------------+-----------------+-------------+------------------+------+-----------------------------------------------------------------+------------------+
| Id | User            | Host            | db          | Command          | Time | State                                                           | Info             |
+----+-----------------+-----------------+-------------+------------------+------+-----------------------------------------------------------------+------------------+
|  5 | event_scheduler | localhost       | NULL        | Daemon           |  467 | Waiting on empty queue                                          | NULL             |
| 13 | vt_dba          | localhost       | NULL        | Sleep            |    1 |                                                                 | NULL             |
| 23 | vt_dba          | localhost       | vt_commerce | Sleep            |  462 |                                                                 | NULL             |
| 24 | vt_app          | localhost       | vt_commerce | Sleep            |  463 |                                                                 | NULL             |
| 26 | vt_allprivs     | localhost       | vt_commerce | Sleep            |  463 |                                                                 | NULL             |
| 27 | vt_repl         | 127.0.0.1:50890 | NULL        | Binlog Dump GTID |  463 | Source has sent all binlog to replica; waiting for more updates | NULL             |
| 28 | vt_repl         | 127.0.0.1:50889 | NULL        | Binlog Dump GTID |  463 | Source has sent all binlog to replica; waiting for more updates | NULL             |
| 32 | vt_app          | localhost       | vt_commerce | Sleep            |   23 |                                                                 | NULL             |
| 39 | vt_app          | localhost       | vt_commerce | Sleep            |  461 |                                                                 | NULL             |
| 42 | root            | localhost       | vt_commerce | Query            |    0 | init                                                            | show processlist |
+----+-----------------+-----------------+-------------+------------------+------+-----------------------------------------------------------------+------------------+
10 rows in set, 1 warning (0.00 sec)

mysql> select variable_value from performance_schema.global_status where variable_name="threads_connected";
+----------------+
| variable_value |
+----------------+
| 9              |
+----------------+
1 row in set (0.00 sec)
```

File: go/test/endtoend/vtcombo/recreate/recreate_test.go [hunk 3]
```diff
@@ -101,13 +101,36 @@ func TestDropAndRecreateWithSameShards(t *testing.T) {
...
+	if err != nil {
+		return 0, err
+	}
+	row := result.Rows[0][0]
+	toInt, err := row.ToInt()
+	if err != nil {
+		return 0, err
+	}
+	return toInt, nil
```
Comment (mattlord) [inline go/test/endtoend/vtcombo/recreate/recreate_test.go:133]: Nit, but you can do all of  this together:
```
return result.Rows[0][0].ToInt()
```

File: go/test/endtoend/vtcombo/recreate/recreate_test.go
Comment (mattlord) [inline go/test/endtoend/vtcombo/recreate/recreate_test.go:1]: 👍 I'm assuming that this fails on main? I think that's what your comment was demonstrating but just triple checking.

File: go/test/endtoend/vtcombo/recreate/recreate_test.go (hunk 1 above)
Comment (brendar) [reply go/test/endtoend/vtcombo/recreate/recreate_test.go:116]: There are zero connections after the `DROP`. I'm assuming the existing connections get closed and cannot be reestablished by the pools. After the `CREATE` the pools reestablish connections and that's when the leak becomes apparent.

File: go/test/endtoend/vtcombo/recreate/recreate_test.go
Comment (brendar) [reply go/test/endtoend/vtcombo/recreate/recreate_test.go:1]: Yes, it fails on main.

File: go/test/endtoend/vtcombo/recreate/recreate_test.go (hunk 1 above)
Comment (mattlord) [reply go/test/endtoend/vtcombo/recreate/recreate_test.go:116]: OK, cool. That's worth a comment then in the test IMO.

**Comment by brendar** [comment]
@mattlord Thanks for the thoughtful review. Comments have been addressed and I've confirmed the test still fails on main.

**Review by mattlord — APPROVED** [review]
Thanks, @brendar !

**Review by deepthi — APPROVED** [review]

**Pull request merged**